agent/consul/internal_endpoint.go

```diff
@@ -340,16 +341,79 @@ func (m *Internal) GatewayServices(args *structs.ServiceSpecificRequest, reply *
 				return fmt.Errorf("service %q is not a configured terminating-gateway or ingress-gateway", args.ServiceName)
 			}
 
-			index, services, err = state.GatewayServices(ws, args.ServiceName, &args.EnterpriseMeta)
+			maxIdx, services, err = state.GatewayServices(ws, args.ServiceName, &args.EnterpriseMeta)
 			if err != nil {
 				return err
 			}
 
+			// Filter out services added by wildcards with mismatched protocols.
+			idx, services, err := filterGatewayServices(ws, args, state, services, args.Datacenter)
```

Comment:
Added filtering in the RPC endpoint, not in the state store, because we want to filter `GatewayServices` based on whether or not the backing service has the same protocol as the defined listener. Since `protocol` can be set in proxy-defaults/service-defaults, maintaining an accurate table in memdb seemed like it could get problematic.

I suppose we could add the filtering to the `state.GatewayServices` method though... 🤔 Thoughts?

Reply:
I think we can do that kind of conflict validation for non-wildcard cases (making sure the upstream services the user sets match the listener's protocol) but for the wildcard case we can't really give an error if the wildcard picks up a service with a different protocol, since that would make wildcards impossible to use unless every service in a DC had the same protocol set. The best way to surface the information about which services are routed to by an ingress gateway is probably the UI/internal endpoint for the gateway services.

The reason we added the filtering here and not in state.GatewayServices was that it was a lot more complex to do it on updates within the state store, since the filtered services could change on an update to any of service registrations/ingress config/service defaults/proxy defaults when a wildcard is used. Maybe there's a simple way to handle the wildcard case though?

Reply:
We moved this filtering logic into the `state.GatewayServices` method, feels better there, I agree.